File: product-brain/src/product_brain/index/rename_tracker.py
from __future__ import annotations

import subprocess
from pathlib import Path


def canonical_paths(repo: Path, paths: list[str]) -> dict[str, str]:
    """Map old paths to their current names at HEAD via `git log --follow`.

    Paths that no longer exist at HEAD are returned as-is (caller marks them
    as deleted). This is a best-effort replay; for repos with many renames,
    consider caching the result of `git log --all --name-status -M` once.
    """
    out: dict[str, str] = {}
    for p in paths:
        if (repo / p).exists():
            out[p] = p
            continue
        try:
            res = subprocess.run(
                ["git", "-C", str(repo), "log", "--follow", "--name-status", "--pretty=", "-1", "--", p],
                capture_output=True, text=True, check=True,
            )
            new_path = p
            for line in res.stdout.splitlines():
                parts = line.split("\t")
                if len(parts) >= 3 and parts[0].startswith("R"):
                    new_path = parts[2]
                    break
            out[p] = new_path
        except subprocess.CalledProcessError:
            out[p] = p
    return out

File: product-brain/src/records/rename-tracker.ts
import { execFileSync } from "node:child_process";
import { existsSync } from "node:fs";
import path from "node:path";

export function canonicalPaths(repo: string, paths: string[]): Record<string, string> {
  const out: Record<string, string> = {};
  for (const p of paths) {
    if (existsSync(path.join(repo, p))) {
      out[p] = p;
      continue;
    }
    let stdout: string;
    try {
      stdout = execFileSync(
        "git",
        ["-C", repo, "log", "--follow", "--name-status", "--pretty=", "-1", "--", p],
        { encoding: "utf8" },
      );
    } catch {
      out[p] = p;
      continue;
    }
    let newPath = p;
    for (const line of stdout.split("\n")) {
      const parts = line.split("\t");
      if (parts.length >= 3 && parts[0]?.startsWith("R")) {
        newPath = parts[2] ?? p;
        break;
      }
    }
    out[p] = newPath;
  }
  return out;
}
